### ISSUE TITLE
maxloc/minloc reductions don't work when indices are strings

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#19108]
### Summary of Problem

Our current implementation of minloc/maxloc take a min() or max() of the idxType, I believe in order to come up with a placeholder/sentinel value for the location in the event that the reduction set is empty (?).  However, when doing a minloc or maxloc over a domain whose idxType is `string` (or anything else that doesn't support `min` or `max`), this results in errors:

```
error: unresolved call 'max(type string)'
$CHPL_HOME/modules/internal/ChapelTuple.chpl:393: note: this candidate did not match: max(type t)
$CHPL_HOME/modules/internal/ChapelTuple.chpl:393: note: because where clause evaluated to false
day14-maxlocstringinds.chpl:6: note: other candidates are:
$CHPL_HOME/modules/standard/Types.chpl:734: note:   max(type t)
$CHPL_HOME/modules/standard/Types.chpl:737: note:   max(type t)
note: and 22 other candidates, use --print-all-candidates to see them
```

It seems as though we need to come up with some other way of generating this placeholder value.

### Steps to Reproduce

**Associated Future Test(s):**
[`test/studies/adventOfCode/2021/bradc/futures/day14-maxlocstringinds.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/studies/adventOfCode/2021/bradc/futures/day14-maxlocstringinds.chpl
          ) #19028 

### Configuration Information

- Output of `chpl --version`: `chpl version 1.26.0 pre-release (56bd177336)`

**Comment by bradcray** [comment]
@vasslitvinov : I think that this is your code—curious if you have ideas here.

**Comment by vasslitvinov** [comment]
For the "loc" component, it needs to be `max(t)` for both minloc and maxloc, because we report the smallest location in both cases. It makes sense that minloc/maxloc reductions require idxType to have an identity element w.r.t. the min() operation. Alas this rules out strings and perhaps other useful things.

To handle the case where idxType is `string`, we could create an "infinity" of a string, which is to be greater than any other string. Do so by playing with the internals of string code/representation.

However this does not support other types that are comparable yet do not have a max value. To support those cases, we have to add a bit into the reduction state that indicates that the value is uninitialized. I believe we used to have that bit and I removed it when I was tidying up these reductions.

Introducing such a bit will have a hopefully mild effect on performance, and is not needed when idxType already has a natural max value.

The only problem now is to define the outcome of such a reduction when the collection being reduced is empty. Because there will be no legal value to place in the "location" component of the result. A couple of prospective solutions:
* the result will include the "initialized" bit in addition to loc and min/max, or
* the reduction will throw an error when the collection is empty.

**Comment by bradcray** [comment]
I don't think we should have an "infinite" string value (at least, simply for this purpose), but would suggest other approaches for idxTypes that don't have min/max values, or comparators, such as:

* reintroducing the uninitialized bit, potentially only for idxTypes like this that need it
* for idxTypes that don't have a well-defined ordering, perhaps we don't make any guarantees about which index is returned if there are multiple max/min values, we just choose one arbitrarily
* for empty reductions, maybe we use the default value of the idxType as the initial value and/or give the user the ability to pass in a sentinel value for the index in such cases as an additional argument to the reduction

**Comment by vasslitvinov** [comment]
Good ideas.

What can be the mechanism for the reduction implementation to determine when the loc/value types have ordering and/or the default value?

One idea is to require uses to invoke a function, not an operator, in such cases. The operator would continue to report an error.

**Comment by bradcray** [comment]
> What can be the mechanism for the reduction implementation to determine when the loc/value types have ordering and/or the default value?

I was imagining using reflection on things like the availability of `max(t)` and/or `<` and/or `hasDefaultValue()` to make such decisions.

**Comment by aconsroe-hpe** [comment]
Relates to:
* #6312
* #6313

**Comment by bradcray** [comment]
Oh, thanks for those cross-references, I had no memory of those issues (and obviously didn't spend much time looking for them).  I'd almost close this issue in favor of #6313 except that this one is more focused and has a future linked to it.  But it's intriguing to me that @ben-albrecht has run into this multiple times (e.g., #6311 and https://github.com/chapel-lang/chapel/issues/6313#issuecomment-570699430).
